File: js/test/serial_tags.test.mjs
/**
 * The TypeScript half of the cross-language tag contract.
 *
 * `testdata/serial_tags.json` is the one table every language answers to, and
 * `testdata/interaction_golden.json` is one interaction as the Python SDK
 * writes it. Between them these tests pin the thing the plugin depends on: a
 * conversation can be held in TypeScript and handed back to a Python backend
 * turn after turn, with the tool calls and results inside each interaction
 * arriving as the very objects that were sent.
 */

import assert from 'node:assert/strict';
import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import test from 'node:test';

import {
  ACTION_STATUS_MIMETYPE,
  ActionMessage,
  CLOSE_STATUS_ATTRIBUTE,
  Chunk,
  ChunkMetadata,
  NodeFragment,
  SerializationRegistry,
  decodeStatusChunk,
  fromChunk,
  isCloseStatusChunk,
  isOk,
  statusToChunk,
  makeTextMessageInteraction,
  makeOllamaCreateChatConfig,
  toChunk,
  valueTag,
  wireValueCodecs,
} from '../dist/index.js';
import * as tags from '../dist/serial_tags.js';

const testdata = (name) =>
  JSON.parse(readFileSync(fileURLToPath(new URL(`../../testdata/${name}`, import.meta.url)), 'utf8'));

const fixtureTags = () => {
  const data = testdata('serial_tags.json');
  const result = [];
  for (const [section, entries] of Object.entries(data)) {
    if (section.startsWith('_')) continue;
    result.push(...Object.values(entries));
  }
  return result;
};

const need = (value) => {
  assert.ok(isOk(value), `expected ok, got ${JSON.stringify(value)}`);
  return value;
};

const goldenChunk = () => {
  const golden = testdata('interaction_golden.json');
  return {
    golden,
    data: Buffer.from(golden.base64, 'base64'),
    chunk: new Chunk({
      data: new Uint8Array(Buffer.from(golden.base64, 'base64')),
      metadata: new ChunkMetadata({ mimetype: golden.mimetype }),
    }),
  };
};

test('every tag in the shared table has a TypeScript constant', () => {
  const constants = new Set(Object.values(tags).filter((value) => typeof value === 'string'));
  const missing = fixtureTags().filter((tag) => !constants.has(tag));
  assert.deepEqual(missing, [], 'tags in testdata/serial_tags.json with no TypeScript constant');
});

test('every TypeScript tag constant is in the shared table', () => {
  const declared = new Set(fixtureTags());
  const extra = Object.entries(tags)
    .filter(([name, value]) => typeof value === 'string' && name.endsWith('_TAG'))
    .map(([, value]) => value)
    .filter((tag) => !declared.has(tag));
  assert.deepEqual(extra, [], 'TypeScript tags absent from testdata/serial_tags.json');
});

test('the runtime and SDK types are registered under their canonical tags', () => {
  const registered = new Set(wireValueCodecs().map((codec) => codec.tag));
  for (const tag of [
    tags.CHUNK_TAG,
    tags.CHUNK_METADATA_TAG,
    tags.NODE_REF_TAG,
    tags.NODE_FRAGMENT_TAG,
    tags.PORT_TAG,
    tags.ACTION_MESSAGE_TAG,
    tags.WIRE_MESSAGE_TAG,
    tags.STATUS_TAG,
    tags.INTERACTION_TAG,
    tags.PEER_TAG,
    tags.ACTION_CONFIG_TAG,
    tags.USAGE_METADATA_TAG,
    tags.INTERACT_WITH_OLLAMA_CONFIG_TAG,
  ]) {
    assert.ok(registered.has(tag), `no wire value codec for ${tag}`);
  }
});

test('a status chunk is the one shape every language writes', () => {
  const fixture = testdata('status_chunk.json');
  assert.equal(ACTION_STATUS_MIMETYPE, fixture.mimetype);
  assert.equal(CLOSE_STATUS_ATTRIBUTE, fixture.close_attribute);

  for (const testCase of fixture.cases) {
    const status = {code: testCase.code, message: testCase.message, details: testCase.details};
    const chunk = need(statusToChunk(status));
    assert.equal(chunk.mimetype, fixture.mimetype, testCase.name);
    assert.equal(Buffer.from(chunk.data).toString('base64'), testCase.base64, testCase.name);
    assert.equal(isCloseStatusChunk(chunk), false, testCase.name);
    const decoded = need(decodeStatusChunk(chunk));
    assert.equal(decoded.status.code, testCase.code, testCase.name);
    assert.equal(decoded.status.message, testCase.message, testCase.name);
    assert.deepEqual(decoded.status.details ?? [], testCase.details, testCase.name);
  }
});

test('a closure marker only adds the shared attribute', () => {
  const fixture = testdata('status_chunk.json');
  const ok = {code: 0, message: ''};
  const plain = need(statusToChunk(ok));
  const marker = need(statusToChunk(ok, true));

  // The marker rides on the metadata, so the payload is the plain status.
  assert.deepEqual(marker.data, plain.data);
  assert.equal(marker.mimetype, plain.mimetype);
  assert.equal(isCloseStatusChunk(marker), true);
  assert.equal(isCloseStatusChunk(plain), false);
  assert.deepEqual(
    [...marker.metadata.attributes].map(([key, value]) => [key, Buffer.from(value).toString()]),
    [[fixture.close_attribute, '1']],
  );
});

test("Python's interaction decodes into real objects, not anonymous fields", async () => {
  const { chunk } = goldenChunk();

  const interaction = need(await fromChunk(chunk));

  assert.equal(valueTag(interaction), tags.INTERACTION_TAG);
  assert.equal(interaction.model, 'golden-model');
  // The point of the exercise: what a turn *did* survives the crossing.
  assert.ok(interaction.content[0] instanceof Chunk);
  assert.equal(interaction.content[0].mimetype, 'application/json');
  assert.ok(interaction.action_calls[0] instanceof ActionMessage);
  assert.equal(interaction.action_calls[0].name, 'rename_symbol');
  assert.ok(interaction.action_inputs.p[0] instanceof NodeFragment);
  assert.equal(interaction.action_inputs.p[0].id, 'n1');
  assert.equal(valueTag(interaction.status), tags.STATUS_TAG);
  assert.equal(valueTag(interaction.usage_metadata), tags.USAGE_METADATA_TAG);
  assert.equal(interaction.usage_metadata.output_tokens, 7);
  assert.equal(valueTag(interaction.action_configs.x), tags.ACTION_CONFIG_TAG);
  assert.ok(interaction.backend_specific_metadata.stop instanceof Uint8Array);
});

test("an interaction handed back is byte-for-byte the one that arrived", async () => {
  const { chunk, data } = goldenChunk();

  const interaction = need(await fromChunk(chunk));
  const reencoded = need(await toChunk(interaction));

  assert.equal(reencoded.mimetype, `application/json;type=${tags.INTERACTION_TAG}`);
  assert.equal(Buffer.from(reencoded.data).toString(), data.toString());
});

test('an interaction built here is tagged for the strict interactions port', async () => {
  const interaction = need(await makeTextMessageInteraction('hello', 'be brief'));

  const chunk = need(await toChunk(interaction));

  assert.equal(chunk.mimetype, `application/json;type=${tags.INTERACTION_TAG}`);
});

test('an interaction built here is the one the Python SDK builds', async () => {
  // This is what a chat client sends on the `interactions` port, and the
  // backend validates it against its own Interaction model -- so `content` and
  // `system_instructions` have to be Chunks, not the bare JSON they hold.
  const golden = testdata('text_message_interaction_golden.json');
  const expected = JSON.parse(Buffer.from(golden.base64, 'base64').toString());

  const interaction = need(
    await makeTextMessageInteraction(golden.text, golden.system_prompt),
  );
  const chunk = need(await toChunk(interaction));
  const actual = JSON.parse(Buffer.from(chunk.data).toString());

  assert.equal(chunk.mimetype, golden.mimetype);
  assert.deepEqual(actual.content, expected.content);
  assert.deepEqual(actual.system_instructions, expected.system_instructions);
  // Every other field this side writes must agree too. It writes fewer of them
  // — Python spells out `status`, `created_at_millis` and `usage_metadata`
  // where this side leaves them to the model's defaults, which is why the two
  // payloads are equivalent rather than identical.
  assert.equal(typeof actual.id, 'string');
  for (const [key, value] of Object.entries(actual)) {
    if (key === 'id') continue;
    assert.deepEqual(value, expected[key], `field ${key}`);
  }
});

test('a provider config is tagged so the unary config port accepts it', async () => {
  const config = need(makeOllamaCreateChatConfig({ temperature: 0.5 }));

  const chunk = need(await toChunk(config));

  assert.equal(
    chunk.mimetype,
    `application/json;type=${tags.INTERACT_WITH_OLLAMA_CONFIG_TAG}`,
  );
});

test('an ordinary object is still ordinary data', async () => {
  // Nothing here may be claimed by a tagged codec on shape alone -- a caller's
  // own {code, message} is data, not a Status.
  const chunk = need(await toChunk({ code: 0, message: 'hi' }));

  assert.equal(chunk.mimetype, 'application/json');
  assert.equal(valueTag(need(await fromChunk(chunk))), null);
});

test('a serialized interaction carries no tags', async () => {
  // The model's own schema says what everything is; nothing repeats it. The
  // chunk's `;type=` names the payload, and from there every nested value sits
  // in a field whose declared type identifies it.
  const keys = (value) => {
    if (Array.isArray(value)) return value.flatMap(keys);
    if (value !== null && typeof value === 'object') {
      return Object.entries(value).flatMap(([key, item]) => [key, ...keys(item)]);
    }
    return [];
  };
  const interaction = need(await makeTextMessageInteraction('hi', 'be brief'));
  const chunk = need(await toChunk(interaction));

  const payload = JSON.parse(new TextDecoder().decode(chunk.data));
  assert.deepEqual(keys(payload).filter((key) => key.startsWith('!')), []);
  assert.deepEqual(payload.content[0], {
    data: 'eyJyb2xlIjoidXNlciIsImNvbnRlbnQiOlt7InR5cGUiOiJ0ZXh0IiwidGV4dCI6ImhpIn1dfQ==',
    metadata: { mimetype: 'application/json' },
  });

  // And it still comes back as Chunks, from the schema alone.
  const decoded = need(await fromChunk(chunk));
  assert.ok(decoded.content[0] instanceof Chunk);
  assert.ok(decoded.system_instructions[0] instanceof Chunk);
  assert.equal(need(await fromChunk(decoded.system_instructions[0])), 'be brief');
});

test('no key in a payload is ever read as a type', async () => {
  // A payload is data. Nothing in it names a type, so nothing is escaped.
  // A11 once wrote a nested value's type as the sole key of a one-entry object
  // ({"!a11.Chunk": {...}}), which meant a caller's own object of that shape
  // had to be escaped on the way out. The type lives in the chunk's metadata
  // now, so these go out byte-for-byte as written.
  for (const value of [
    { [`!${tags.CHUNK_TAG}`]: 'not one' },
    { '!whatever': [1, 2] },
    { '!a': 1, b: 2 },
  ]) {
    const chunk = need(await toChunk(value));
    assert.deepEqual(JSON.parse(new TextDecoder().decode(chunk.data)), value);
    assert.deepEqual(need(await fromChunk(chunk)), value);
  }
});

test('an unloadable type tag still yields the payload', async () => {
  // A peer that never imported the naming module still holds valid JSON, and
  // is entitled to read it as such rather than be refused.
  const chunk = new Chunk({
    data: new TextEncoder().encode('{"anything":1}'),
    metadata: new ChunkMetadata({ mimetype: 'application/json;type=some.other.Model' }),
  });

  assert.deepEqual(need(await fromChunk(chunk)), { anything: 1 });
});

test('a registry built before an SDK import still sees its codecs', async () => {
  // The SDK registers on import, routinely after a registry was constructed.
  const registry = new SerializationRegistry({ registerDefaults: true });
  const interaction = need(await makeTextMessageInteraction('hi'));

  const chunk = need(await registry.toChunk(interaction));

  assert.equal(chunk.mimetype, `application/json;type=${tags.INTERACTION_TAG}`);
});

File: kotlin/src/test/kotlin/a11/SerialTagsTest.kt
package a11

import a11.sdk.Interaction
import a11.sdk.ensureInteractionCodec
import a11.sdk.makeTextMessageInteraction
import java.nio.file.Files
import java.nio.file.Path
import java.util.Base64
import kotlin.test.Test
import kotlin.test.assertContentEquals
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

/**
 * The Kotlin half of the cross-language tag contract.
 *
 * `testdata/serial_tags.json` is the one table every language answers to, and
 * `testdata/interaction_golden.json` is
 * one interaction as the Python SDK writes
 * it. Between them these tests pin what a Kotlin client depends on when it
 * holds a conversation and hands it back to a Python backend turn after turn:
 * the tool calls and results inside each interaction must arrive as the objects
 * that were sent, and go back out as the payload that was received.
 */
class SerialTagsTest {
    private fun <T> ok(value: StatusOr<T>): T = when (value) {
        is Ok<T> -> value.value
        is Status -> throw AssertionError("expected ok, got $value")
    }

    private fun testdata(name: String): String {
        // The Gradle build runs in kotlin/; the shared fixtures sit beside it.
        var directory: Path? = Path.of("").toAbsolutePath()
        while (directory != null && !Files.exists(directory.resolve("testdata/$name"))) {
            directory = directory.parent
        }
        assertNotNull(directory, "could not find testdata/$name above the working directory")
        return Files.readString(directory.resolve("testdata/$name"))
    }

    private fun fixtureTags(): List<String> {
        val parsed = ok(A11Json.parse(testdata("serial_tags.json"))) as Map<*, *>
        return parsed.entries
            // `media_types` is the other half of the fixture and holds media
            // types, not tags; see `media types match the fixture`.
            .filter { !it.key.toString().startsWith("_") && it.key != "media_types" }
            .flatMap { (it.value as Map<*, *>).values.map { tag -> tag.toString() } }
    }

    private fun fixtureMediaTypes(): Map<*, *> {
        val parsed = ok(A11Json.parse(testdata("serial_tags.json"))) as Map<*, *>
        return parsed["media_types"] as Map<*, *>
    }

    private fun goldenChunk(): Pair<Chunk, ByteArray> {
        val golden = ok(A11Json.parse(testdata("interaction_golden.json"))) as Map<*, *>
        val data = Base64.getDecoder().decode(golden["base64"].toString())
        return Chunk(
            metadata = ChunkMetadata(mimetype = golden["mimetype"].toString()),
            data = data,
        ) to data
    }

    private fun kotlinTags(): List<String> = listOf(
        SerialTags.CHUNK_METADATA, SerialTags.CHUNK, SerialTags.NODE_REF,
        SerialTags.NODE_FRAGMENT, SerialTags.PORT, SerialTags.ACTION_MESSAGE,
        SerialTags.WIRE_MESSAGE, SerialTags.STATUS, SerialTags.TIME, SerialTags.DURATION,
        SerialTags.INTERACTION, SerialTags.PEER, SerialTags.ACTION_CONFIG,
        SerialTags.USAGE_METADATA, SerialTags.INTERACT_WITH_CLAUDE_CONFIG,
        SerialTags.INTERACT_WITH_CLAUDE_CODE_CONFIG,
        SerialTags.INTERACT_WITH_GEMINI_CONFIG, SerialTags.INTERACT_WITH_OLLAMA_CONFIG,
        SerialTags.INTERACT_WITH_GPT_CONFIG, SerialTags.INTERACT_WITH_CODEX_CONFIG,
        SerialTags.INTERACT_WITH_VLLM_CONFIG,
        SerialTags.INTERACT_WITH_GEMMA_CONFIG, SerialTags.AUDIO_BUFFER,
        SerialTags.AUDIO_INPUT_OPTIONS, SerialTags.SPEECH_RECOGNIZER_OPTIONS,
        SerialTags.AUDIO_DEVICE_INFO, SerialTags.AUDIO_CONTROL_EVENT,
        SerialTags.AUDIO_CAPTURE_EVENT, SerialTags.TRANSCRIPTION_EVENT,
    )

    @Test
    fun kotlinConstantsMatchTheSharedTable() {
        assertEquals(fixtureTags().sorted(), kotlinTags().sorted())
    }

    @Test
    fun aStatusChunkIsTheOneShapeEveryLanguageWrites() {
        val fixture = ok(A11Json.parse(testdata("status_chunk.json"))) as Map<*, *>
        assertEquals(fixture["mimetype"], ACTION_STATUS_MIMETYPE)
        assertEquals(fixture["close_attribute"], CLOSE_STATUS_ATTRIBUTE)

        for (entry in fixture["cases"] as List<*>) {
            val case = entry as Map<*, *>
            val name = case["name"].toString()
            @Suppress("UNCHECKED_CAST")
            val details = (case["details"] as List<*>).map { it as Map<String, Any?> }
            val code = StatusCode.fromValue((case["code"] as Number).toInt())
            assertNotNull(code, name)
            val status = Status(code, case["message"].toString(), details)

            val chunk = ok(statusToChunk(status))
            assertEquals(fixture["mimetype"], chunk.mimetype, name)
            assertEquals(case["base64"], Base64.getEncoder().encodeToString(chunk.data), name)
            assertTrue(!isCloseStatusChunk(chunk), name)

            val decoded = ok(decodeStatusChunk(chunk)).status
            assertEquals(status.code, decoded.code, name)
            assertEquals(status.message, decoded.message, name)
            assertEquals(details, decoded.details, name)
        }
    }

    @Test
    fun aClosureMarkerOnlyAddsTheSharedAttribute() {
        val fixture = ok(A11Json.parse(testdata("status_chunk.json"))) as Map<*, *>
        val plain = ok(statusToChunk(Status.ok()))
        val marker = ok(statusToChunk(Status.ok(), closing = true))

        // The marker rides on the metadata, so the payload is the plain status.
        assertContentEquals(plain.data, marker.data)
        assertEquals(plain.mimetype, marker.mimetype)
        assertTrue(isCloseStatusChunk(marker))
        assertTrue(!isCloseStatusChunk(plain))
        val attributes = assertNotNull(marker.metadata).attributes
        assertEquals(setOf(fixture["close_attribute"]), attributes.keys.toSet())
        assertContentEquals("1".toByteArray(), attributes[fixture["close_attribute"]])
    }

    @Test
    fun theRuntimeTypesAreRegisteredUnderTheirCanonicalTags() {
        ensureCoreWireValues()
        val registered = allWireValueCodecs().map { it.tag }.toSet()
        for (tag in listOf(
            SerialTags.CHUNK, SerialTags.CHUNK_METADATA, SerialTags.NODE_REF,
            SerialTags.NODE_FRAGMENT, SerialTags.PORT, SerialTags.ACTION_MESSAGE,
            SerialTags.WIRE_MESSAGE, SerialTags.STATUS,
        )) {
            assertTrue(registered.contains(tag), "no wire value codec for $tag")
        }
    }

    @Test
    fun pythonsInteractionDecodesIntoRealObjects() {
        ensureInteractionCodec()
        val registry = SerializationRegistry(registerDefaults = true)
        val (chunk, _) = goldenChunk()

        val interaction = ok(registry.fromChunk(chunk)) as Interaction

        assertEquals("golden-model", interaction["model"])
        // Turn metadata survives the language boundary.
        val content = interaction["content"] as List<*>
        assertTrue(content[0] is Chunk)
        assertEquals("application/json", (content[0] as Chunk).mimetype)
        val calls = interaction["action_calls"] as List<*>
        assertTrue(calls[0] is ActionMessage)
        assertEquals("rename_symbol", (calls[0] as ActionMessage).name)
        val inputs = (interaction["action_inputs"] as Map<*, *>)["p"] as List<*>
        assertTrue(inputs[0] is NodeFragment)
        assertEquals("n1", (inputs[0] as NodeFragment).id)
        assertTrue(interaction["status"] is Status)
        assertTrue(interaction["backend_specific_metadata"] is Map<*, *>)
    }

    @Test
    fun anInteractionHandedBackIsTheOneThatArrived() {
        ensureInteractionCodec()
        val registry = SerializationRegistry(registerDefaults = true)
        val (chunk, data) = goldenChunk()

        val interaction = ok(registry.fromChunk(chunk))
        val reencoded = ok(registry.toChunk(interaction))

        assertEquals("application/json;type=${SerialTags.INTERACTION}", reencoded.mimetype)
        assertContentEquals(data, reencoded.data)
    }

    @Test
    fun anInteractionBuiltHereIsTaggedForTheStrictPort() {
        ensureInteractionCodec()
        val registry = SerializationRegistry(registerDefaults = true)

        val chunk = ok(registry.toChunk(ok(makeTextMessageInteraction("hello"))))

        assertEquals("application/json;type=${SerialTags.INTERACTION}", chunk.mimetype)
    }

    @Test
    fun anInteractionBuiltHereIsTheOneThePythonSdkBuilds() {
        // This is what a chat client sends on the `interactions` port, and the
        // backend validates it against its own Interaction model -- so
        // `content` and `system_instructions` have to be Chunks, not the bare
        // values they hold.
        ensureInteractionCodec()
        val registry = SerializationRegistry(registerDefaults = true)
        val golden = ok(A11Json.parse(testdata("text_message_interaction_golden.json"))) as Map<*, *>
        val expected = ok(
            A11Json.parse(String(Base64.getDecoder().decode(golden["base64"].toString()))),
        ) as Map<*, *>

        val interaction = ok(
            makeTextMessageInteraction(
                golden["text"].toString(),
                golden["system_prompt"].toString(),
                registry = registry,
            ),
        )
        val chunk = ok(registry.toChunk(interaction))
        val actual = ok(A11Json.parse(String(chunk.data))) as Map<*, *>

        assertEquals(golden["mimetype"].toString(), chunk.mimetype)
        assertEquals(expected["content"], actual["content"])
        assertEquals(expected["system_instructions"], actual["system_instructions"])
    }

    @Test
    fun anOrdinaryMapIsStillOrdinaryData() {
        val registry = SerializationRegistry(registerDefaults = true)

        val chunk = ok(registry.toChunk(linkedMapOf<String, Any?>("code" to 0L, "message" to "hi")))

        // JSON already says this is an object; the mimetype does not repeat it.
        assertEquals("application/json", chunk.mimetype)
    }

    @Test
    fun aSerializedInteractionCarriesNoTags() {
        // The model's own shape says what everything is; nothing repeats it.
        // The chunk's `;type=` names the payload, and from there every nested
        // value sits in a field whose declared type identifies it.
        ensureInteractionCodec()
        val registry = SerializationRegistry(registerDefaults = true)
        fun keys(value: Any?): List<String> = when (value) {
            is Map<*, *> -> value.entries.flatMap { listOf(it.key.toString()) + keys(it.value) }
            is List<*> -> value.flatMap { keys(it) }
            else -> emptyList()
        }
        val interaction = ok(makeTextMessageInteraction("hi", "be brief"))

        val chunk = ok(registry.toChunk(interaction))

        val payload = ok(A11Json.parse(String(chunk.data)))
        assertEquals(emptyList(), keys(payload).filter { it.startsWith("!") })

        // And it still comes back as Chunks, from the shape alone.
        val decoded = ok(registry.fromChunk(chunk)) as Interaction
        assertTrue((decoded["content"] as List<*>)[0] is Chunk)
        val instructions = decoded["system_instructions"] as List<*>
        assertEquals("be brief", ok(registry.fromChunk(instructions[0] as Chunk)))
    }

    @Test
    fun noKeyInAPayloadIsEverReadAsAType() {
        // A payload is data. Nothing in it names a type, so nothing is escaped.
        // A11 once wrote a nested value's type as the sole key of a one-entry
        // object ({"!a11.Chunk": {...}}), which meant a caller's own map of
        // that shape had to be escaped on the way out. The type lives in the
        // chunk's metadata now, so these go out byte-for-byte as written.
        val registry = SerializationRegistry(registerDefaults = true)
        for (value in listOf(
            linkedMapOf<String, Any?>("!${SerialTags.CHUNK}" to "not one"),
            linkedMapOf<String, Any?>("!whatever" to listOf(1L, 2L)),
            linkedMapOf<String, Any?>("!a" to 1L, "b" to 2L),
        )) {
            assertEquals(value, ok(registry.fromChunk(ok(registry.toChunk(value)))))
        }
        val plain = linkedMapOf<String, Any?>("!a" to 1L, "b" to 2L)
        assertEquals("""{"!a":1,"b":2}""", ok(utf8Decode(ok(registry.toChunk(plain)).data)))
    }

    @Test
    fun anUnloadableTypeTagStillYieldsThePayload() {
        // A peer that never loaded the naming module still holds valid JSON,
        // and is entitled to read it as such rather than be refused.
        val registry = SerializationRegistry(registerDefaults = true)
        val chunk = Chunk(
            metadata = ChunkMetadata(mimetype = "application/json;type=some.other.Model"),
            data = utf8Encode("""{"anything":1}"""),
        )

        assertEquals(
            linkedMapOf<String, Any?>("anything" to 1L),
            ok(registry.fromChunk(chunk)),
        )
    }

    @Test
    fun mediaTypesMatchTheFixture() {
        // Pinned across languages exactly as the tags are. `text` and `bytes`
        // are the ones that matter: they are the defaults for a String and a
        // ByteArray, and a chunk using either carries no `type` parameter, so
        // the media type alone is what a peer has to go on.
        val media = fixtureMediaTypes()

        assertEquals(media["json"], JSON_MIMETYPE)
        assertEquals(media["msgpack"], MSGPACK_MIMETYPE)
        assertEquals(media["text"], TEXT_MIMETYPE)
        // This side's constant keeps its established name; what is pinned is
        // the fixture key, not the identifier each language spells it with.
        assertEquals(media["bytes"], OCTET_STREAM_MIMETYPE)
    }

    @Test
    fun aStringAndBytesTravelAsThemselves() {
        val registry = SerializationRegistry(registerDefaults = true)
        val media = fixtureMediaTypes()

        val text = ok(registry.toChunk("value"))
        val bytes = ok(registry.toChunk(byteArrayOf(0, -1)))

        assertEquals(media["text"], text.mimetype)
        assertEquals(media["bytes"], bytes.mimetype)
        // No framing: the bytes on the wire are the value.
        assertEquals("value", ok(utf8Decode(text.data)))
        assertContentEquals(byteArrayOf(0, -1), bytes.data)
        assertEquals("value", ok(registry.fromChunk(text)))
        assertContentEquals(byteArrayOf(0, -1), ok(registry.fromChunk(bytes)) as ByteArray)
    }
}
